File: backend/app/schemas/application_inventory.py
from __future__ import annotations

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field


class ApplicationInventoryItemUpdateRequest(BaseModel):
    application_number_apm: str | None = None
    parent_application_name: str | None = None
    assignment_group: str | None = None
    assignment_group_owner: str | None = None
    application_owner: str | None = None
    business_service_ci_name: str | None = None
    support_lead: str | None = None
    functional_track: str | None = None
    ams_owner: str | None = None
    supported_by_vendor: str | None = None
    hosting_env: str | None = None
    global_application: str | None = None
    scope_status: str | None = None
    lifecycle_stage_status: str | None = None
    lifecycle_current: str | None = None
    lifecycle_1_to_3_years: str | None = None
    lifecycle_3_to_5_years: str | None = None
    active: bool | None = None
    active_users: int | None = None
    avg_monthly_ticket_volume_6m: float | None = None
    tickets_per_user_per_month: float | None = None
    cmdb_payload: dict[str, Any] | None = None


class ApplicationInventoryItemResponse(BaseModel):
    id: UUID
    project_id: UUID
    application_number_apm: str | None
    parent_application_name: str | None
    assignment_group: str | None
    assignment_group_owner: str | None
    application_owner: str | None
    business_service_ci_name: str
    support_lead: str | None
    functional_track: str | None
    ams_owner: str | None
    supported_by_vendor: str | None
    hosting_env: str | None
    global_application: str | None
    scope_status: str
    lifecycle_stage_status: str | None
    lifecycle_current: str | None
    lifecycle_1_to_3_years: str | None
    lifecycle_3_to_5_years: str | None
    active: bool | None
    active_users: int | None
    avg_monthly_ticket_volume_6m: float | None
    tickets_per_user_per_month: float | None
    cmdb_payload: dict[str, Any] | None
    source_filename: str | None
    source_row_number: int | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ApplicationInventoryUploadResponse(BaseModel):
    project_id: UUID
    total_rows: int
    inserted_count: int
    updated_count: int
    skipped_count: int
    error_count: int
    warning_count: int
    errors: list[str] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)
    distinct_business_service_count: int
    distinct_parent_application_count: int
    distinct_assignment_group_count: int
    distinct_application_owner_count: int
    distinct_support_lead_count: int
    distinct_functional_track_count: int
    distinct_ams_owner_count: int
    distinct_supported_vendor_count: int


class ApplicationInventoryEnrichRequest(BaseModel):
    project_id: UUID
    replace_existing: bool = True


class ApplicationTicketUserMetricsRecomputeResponse(BaseModel):
    project_id: UUID
    inventory_count: int
    active_users_count: int
    metrics_updated_count: int
    window_start: datetime
    window_end: datetime


class ValueCountResponse(BaseModel):
    value: str
    count: int


class ApplicationInventoryEnrichmentSummaryResponse(BaseModel):
    project_id: UUID
    total_tickets: int
    matched_tickets: int
    unmatched_tickets: int
    updated_tickets: int
    match_rate_pct: float | None
    matched_by_business_service_count: int
    matched_by_application_count: int
    unmatched_business_service_count: int
    distinct_ticket_business_service_count: int
    distinct_inventory_business_service_count: int
    top_unmatched_business_services: list[ValueCountResponse]
    top_unmatched_applications: list[ValueCountResponse]
    top_unmatched_assignment_groups: list[ValueCountResponse]


class UnmatchedBusinessServiceResponse(BaseModel):
    business_service: str
    ticket_count: int
    assignment_group_count: int
    sample_assignment_groups: list[str]
    sample_ticket_numbers: list[str]


class UnmatchedBusinessServicesResponse(BaseModel):
    project_id: UUID
    distinct_ticket_business_service_count: int
    distinct_inventory_business_service_count: int
    matched_business_service_count: int
    unmatched_business_service_count: int
    business_service_coverage_pct: float | None
    rows: list[UnmatchedBusinessServiceResponse]


class ApplicationInventoryFilterValuesResponse(BaseModel):
    application_owners: list[str]
    support_leads: list[str]
    functional_tracks: list[str]
    ams_owners: list[str]
    supported_by_vendors: list[str]
    hosting_envs: list[str]
    parent_application_names: list[str]
    business_service_ci_names: list[str]
    assignment_groups: list[str]


class ScopeSummaryValueCountResponse(BaseModel):
    value: str
    count: int


class ScopeSummaryResponse(BaseModel):
    project_id: UUID
    in_scope_tickets: int
    out_of_scope_tickets: int
    total_classified_tickets: int
    in_scope_pct: float | None
    out_of_scope_pct: float | None
    distinct_in_scope_assignment_groups: int
    distinct_out_of_scope_assignment_groups: int
    top_out_of_scope_assignment_groups: list[ScopeSummaryValueCountResponse]
    top_out_of_scope_business_services: list[ScopeSummaryValueCountResponse]

File: frontend/src/api/applicationInventory.ts
import { requestJson } from "./client";

export type ApplicationInventoryItem = {
  id: string;
  project_id: string;
  application_number_apm: string | null;
  parent_application_name: string | null;
  assignment_group: string | null;
  assignment_group_owner: string | null;
  application_owner: string | null;
  business_service_ci_name: string;
  support_lead: string | null;
  functional_track: string | null;
  ams_owner: string | null;
  supported_by_vendor: string | null;
  hosting_env: string | null;
  global_application: string | null;
  scope_status: string;
  lifecycle_stage_status: string | null;
  lifecycle_current: string | null;
  lifecycle_1_to_3_years: string | null;
  lifecycle_3_to_5_years: string | null;
  active: boolean | null;
  active_users: number | null;
  cmdb_payload: Record<string, unknown> | null;
  source_filename: string | null;
  source_row_number: number | null;
  created_at: string;
  updated_at: string;
};

export type ApplicationInventoryUploadResponse = {
  project_id: string;
  total_rows: number;
  inserted_count: number;
  updated_count: number;
  skipped_count: number;
  error_count: number;
  warning_count: number;
  errors: string[];
  warnings: string[];
  distinct_business_service_count: number;
  distinct_parent_application_count: number;
  distinct_assignment_group_count: number;
  distinct_application_owner_count: number;
  distinct_support_lead_count: number;
  distinct_functional_track_count: number;
  distinct_ams_owner_count: number;
  distinct_supported_vendor_count: number;
};

export type ValueCount = {
  value: string;
  count: number;
};

export type ApplicationInventoryEnrichmentSummary = {
  project_id: string;
  total_tickets: number;
  matched_tickets: number;
  unmatched_tickets: number;
  updated_tickets: number;
  match_rate_pct: number | null;
  matched_by_business_service_count: number;
  matched_by_application_count: number;
  unmatched_business_service_count: number;
  distinct_ticket_business_service_count: number;
  distinct_inventory_business_service_count: number;
  top_unmatched_business_services: ValueCount[];
  top_unmatched_applications: ValueCount[];
  top_unmatched_assignment_groups: ValueCount[];
};

export type UnmatchedBusinessServiceRow = {
  business_service: string;
  ticket_count: number;
  assignment_group_count: number;
  sample_assignment_groups: string[];
  sample_ticket_numbers: string[];
};

export type UnmatchedBusinessServicesResponse = {
  project_id: string;
  distinct_ticket_business_service_count: number;
  distinct_inventory_business_service_count: number;
  matched_business_service_count: number;
  unmatched_business_service_count: number;
  business_service_coverage_pct: number | null;
  rows: UnmatchedBusinessServiceRow[];
};

export type ApplicationInventoryFilterValues = {
  application_owners: string[];
  support_leads: string[];
  functional_tracks: string[];
  ams_owners: string[];
  supported_by_vendors: string[];
  hosting_envs: string[];
  parent_application_names: string[];
  business_service_ci_names: string[];
  assignment_groups: string[];
};

export type ScopeSummaryValueCount = {
  value: string;
  count: number;
};

export type ScopeSummary = {
  project_id: string;
  in_scope_tickets: number;
  out_of_scope_tickets: number;
  total_classified_tickets: number;
  in_scope_pct: number | null;
  out_of_scope_pct: number | null;
  distinct_in_scope_assignment_groups: number;
  distinct_out_of_scope_assignment_groups: number;
  top_out_of_scope_assignment_groups: ScopeSummaryValueCount[];
  top_out_of_scope_business_services: ScopeSummaryValueCount[];
};

export function listApplicationInventory(projectId: string): Promise<ApplicationInventoryItem[]> {
  const query = new URLSearchParams({ project_id: projectId.trim() });
  return requestJson<ApplicationInventoryItem[]>(`/application-inventory?${query.toString()}`);
}

export function uploadApplicationInventoryFile(
  projectId: string,
  file: File
): Promise<ApplicationInventoryUploadResponse> {
  const formData = new FormData();
  formData.append("project_id", projectId.trim());
  formData.append("file", file);

  return requestJson<ApplicationInventoryUploadResponse>("/application-inventory/upload", {
    method: "POST",
    body: formData,
  });
}

export function enrichApplicationInventory(
  projectId: string,
  replaceExisting: boolean
): Promise<ApplicationInventoryEnrichmentSummary> {
  return requestJson<ApplicationInventoryEnrichmentSummary>(
    "/application-inventory/enrich-tickets",
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        project_id: projectId.trim(),
        replace_existing: replaceExisting,
      }),
    }
  );
}

export function getApplicationInventorySummary(
  projectId: string
): Promise<ApplicationInventoryEnrichmentSummary> {
  const query = new URLSearchParams({ project_id: projectId.trim() });
  return requestJson<ApplicationInventoryEnrichmentSummary>(
    `/application-inventory/enrichment-summary?${query.toString()}`
  );
}

export function getUnmatchedBusinessServices(
  projectId: string,
  limit = 50
): Promise<UnmatchedBusinessServicesResponse> {
  const query = new URLSearchParams({
    project_id: projectId.trim(),
    limit: String(limit),
  });
  return requestJson<UnmatchedBusinessServicesResponse>(
    `/application-inventory/unmatched-business-services?${query.toString()}`
  );
}

export function getApplicationInventoryFilterValues(
  projectId: string
): Promise<ApplicationInventoryFilterValues> {
  const query = new URLSearchParams({ project_id: projectId.trim() });
  return requestJson<ApplicationInventoryFilterValues>(
    `/application-inventory/filter-values?${query.toString()}`
  );
}

export function getScopeSummary(projectId: string): Promise<ScopeSummary> {
  const query = new URLSearchParams({ project_id: projectId.trim() });
  return requestJson<ScopeSummary>(`/application-inventory/scope-summary?${query.toString()}`);
}
